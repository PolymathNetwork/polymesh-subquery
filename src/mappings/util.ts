import { Codec } from "@polkadot/types/types";
import { u8aToString } from "@polkadot/util";
/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);

export const capitalizeFirstLetter = (str: string): string =>
  str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s.replace(/\0/g, "");

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (
  text: string,
  exitOnFirst = false
): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case "(":
      case "{":
      case "<": {
        nestedLevel++;
        break;
      }
      case ")":
      case "}":
      case ">": {
        nestedLevel--;
        break;
      }
      case ",": {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(
      `No top level comma found in ${text}, it probably isn't a map`
    );
  }
  return commas;
};

export const getOrDefault = <K, V>(
  map: Map<K, V>,
  key: K,
  getDefault: () => V
): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};
