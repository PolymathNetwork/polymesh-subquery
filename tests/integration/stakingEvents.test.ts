import { gql } from "@apollo/client/core";
import { getApolloClient } from "../util";
const { query } = getApolloClient();

describe("stakingEvents", () => {
  test("should return staking events", async () => {
    const q = {
      query: gql`
        query {
          stakingEvents {
            nodes {
              stakingEventId
              date
              identityId
              stashAccount
              amount
              nominatedValidators
            }
            totalCount
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
  test("should return nomination events", async () => {
    const q = {
      query: gql`
        query {
          stakingEvents(filter: { stakingEventId: { equalTo: "Nominated" } }) {
            nodes {
              stakingEventId
              date
              identityId
              stashAccount
              amount
              nominatedValidators
            }
            totalCount
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
  test("should filter by account", async () => {
    const q = {
      query: gql`
        query {
          stakingEvents(
            filter: {
              stashAccount: {
                equalTo: "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121"
              }
            }
          ) {
            nodes {
              stakingEventId
              date
              identityId
              stashAccount
              amount
              nominatedValidators
            }
            totalCount
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
});
