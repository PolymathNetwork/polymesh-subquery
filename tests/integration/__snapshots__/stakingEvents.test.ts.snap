// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stakingEvents should filter by account 1`] = `
Object {
  "stakingEvents": Object {
    "__typename": "StakingEventsConnection",
    "nodes": Array [
      Object {
        "__typename": "StakingEvent",
        "amount": "1000000",
        "date": "2021-10-06T12:35:22",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": null,
        "stakingEventId": "Bonded",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": null,
        "date": "2021-10-06T12:35:22.5",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": Array [
          "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
        ],
        "stakingEventId": "Nominated",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": "100",
        "date": "2021-10-06T12:38:10",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": null,
        "stakingEventId": "Unbonded",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": null,
        "date": "2021-10-06T12:38:10.5",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        ],
        "stakingEventId": "Nominated",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
    ],
    "totalCount": 4,
  },
}
`;

exports[`stakingEvents should return nomination events 1`] = `
Object {
  "stakingEvents": Object {
    "__typename": "StakingEventsConnection",
    "nodes": Array [
      Object {
        "__typename": "StakingEvent",
        "amount": null,
        "date": "2021-10-06T12:35:22.5",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": Array [
          "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
        ],
        "stakingEventId": "Nominated",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": null,
        "date": "2021-10-06T12:38:10.5",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        ],
        "stakingEventId": "Nominated",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
    ],
    "totalCount": 2,
  },
}
`;

exports[`stakingEvents should return staking events 1`] = `
Object {
  "stakingEvents": Object {
    "__typename": "StakingEventsConnection",
    "nodes": Array [
      Object {
        "__typename": "StakingEvent",
        "amount": "1000000",
        "date": "2021-10-06T12:34:02.5",
        "identityId": "0x686a05bf46a048e84adf81fb5ba45fbd01a77877ccfe2a638f0215976fafeafd",
        "nominatedValidators": null,
        "stakingEventId": "Bonded",
        "stashAccount": "0x58bde49478e69c60e80b6c80f10a78b8c360489cb445d535fd091a2c3dbfb56b",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": "1000000",
        "date": "2021-10-06T12:35:22",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": null,
        "stakingEventId": "Bonded",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": null,
        "date": "2021-10-06T12:35:22.5",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": Array [
          "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
        ],
        "stakingEventId": "Nominated",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": "76179499",
        "date": "2021-10-06T12:38:06",
        "identityId": "0x0400000000000000000000000000000000000000000000000000000000000000",
        "nominatedValidators": null,
        "stakingEventId": "Reward",
        "stashAccount": "0xfe65717dad0447d715f660a0a58411de509b42e6efb8375f562f58a554d5860e",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": "100",
        "date": "2021-10-06T12:38:10",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": null,
        "stakingEventId": "Unbonded",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
      Object {
        "__typename": "StakingEvent",
        "amount": null,
        "date": "2021-10-06T12:38:10.5",
        "identityId": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "nominatedValidators": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        ],
        "stakingEventId": "Nominated",
        "stashAccount": "0x301ae89b2d71df51a78bfa5700385a7612bd34cf0a482407eccdafc17ef75121",
      },
    ],
    "totalCount": 6,
  },
}
`;
